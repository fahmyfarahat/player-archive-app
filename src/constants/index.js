export const API_URL = "/api";
export const ERROR_MESSAGE = "Player is not available.";
export const HTTP_HEADER_OPTIONS = { accept: "application/json" };

export const FETCH_PLAYER_REQUEST = "FETCH_PLAYER_REQUEST";
export const FETCH_PLAYER_SUCCESS = "FETCH_PLAYER_SUCCESS";
export const FETCH_PLAYER_FAILURE = "FETCH_PLAYER_FAILURE";
export const FETCH_PROFILE_REQUEST = "FETCH_PROFILE_REQUEST";
export const FETCH_PROFILE_SUCCESS = "FETCH_PROFILE_SUCCESS";
export const FETCH_PROFILE_FAILURE = "FETCH_PROFILE_FAILURE";

export const STATS_SCHEMA = {
  attack : ["minutesPerGoal", "goals", "assists", "shotsOnTarget", "totalShots"],
  defense: ["blocks", "totalTackles", "tacklesWon", "interceptions", "clearances"],
  passes : ["totalPasses", "totalCrosses", "passingAccuracy", "successfulCrosses"],
  cards : ["successfulCrosses", "yellowCards", "foulsConceded"],
};