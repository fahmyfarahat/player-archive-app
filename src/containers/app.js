import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { fetchPlayerData, fetchProfileData } from "../actions";
import { ERROR_MESSAGE } from "../constants";
import { 
  isNotEmpty,
  showErrorView, 
  showPlayerView 
} from "../utilities";
import { 
  SearchInputForm, 
  PlayerProfileView, 
  ErrorMessageView 
} from "../components";

export class App extends Component {

  handleOnSubmit = inputValue => {
    const { getPlayerData, getProfileData } = this.props;

    getPlayerData(inputValue)
      .then(() => {
        const { player } = this.props;
        if (isNotEmpty(player.data) && player.data.active) {
          getProfileData(player.data["profile-id"]);
        }
      });
  };

  render() {
    const { player, profile } = this.props;

    return (
      <div>
        <SearchInputForm
          onSubmit={this.handleOnSubmit}
          buttonText="go" 
          placeholder="Type Player's ID" 
          name="searchInput"
        />
        {showPlayerView({ player, profile }) && (<PlayerProfileView profile={profile.data} />)}
        {showErrorView({ player, profile }) && (<ErrorMessageView className="error-message" message={ERROR_MESSAGE} />)}
      </div>
    )
  }
}

const mapStateToProps = ({ player, profile }) => ({ 
  player,
  profile
 });

const mapDispatchToProps = dispatch => ({
  getPlayerData : id => dispatch(fetchPlayerData(id)),
  getProfileData : id => dispatch(fetchProfileData(id))
});

App.propTypes = {
  getPlayerData: PropTypes.func.isRequired,
  getProfileData: PropTypes.func.isRequired,
  profile: PropTypes.object.isRequired,
  player: PropTypes.shape({
    data: PropTypes.object.isRequired,
    error: PropTypes.bool.isRequired,
    loading: PropTypes.bool.isRequired
  }),
};

export default connect(mapStateToProps, mapDispatchToProps)(App);