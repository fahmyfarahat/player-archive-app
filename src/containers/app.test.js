import React from "react";
import { shallow, mount } from "enzyme";
import { App } from "./app";
import { 
  PlayerProfileView, 
  ErrorMessageView 
} from "../components";
import { ERROR_MESSAGE } from "../constants";
import mockFabioProfile from "../../json/profile-333.json"
import toJson from "enzyme-to-json";

describe("App - general", () => {
  let props;
  beforeEach(()=> {
    const initData = {
      data: {},
      error: false,
      loading: false,
    }
    props = {
      getPlayerData: jest.fn(),
      getProfileData: jest.fn(),
      player: initData,
      profile: initData
    };
  });

  it("should render without crashing", () => {
    const wrapper = shallow(<App {...props} />);
    expect(wrapper).toHaveLength(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render error message view if fetch Player Failure", async () => {
    props.player = {
      ...props.player,
      error: true
    };
    const wrapper = shallow(<App {...props} />);
    expect(wrapper.find(ErrorMessageView).props().message).toEqual(ERROR_MESSAGE)
  });

  it("should render error message view if player active=false", async () => {
    props.player.data = {
      "active": false,
      "id": "giorgio",
      "profile-id": "profile-333.json"
    };
    const wrapper = shallow(<App {...props} />);
    expect(wrapper.find(ErrorMessageView).props().message).toEqual(ERROR_MESSAGE)
  });

  it("should render player profile view if player active=true", async () => {
    props.player = {
      ...props.player,
      data: {
        "active": true,
        "id": "fabio",
        "profile-id": "profile-111.json"
      }
    };
    props.profile = {
      ...props.profile,
      data: mockFabioProfile
    }

    const wrapper = mount(<App {...props} />);
    expect(wrapper.find(PlayerProfileView)).toHaveLength(1);
  });
});
