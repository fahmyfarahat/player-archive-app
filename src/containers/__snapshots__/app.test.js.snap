// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App - general should render without crashing 1`] = `
<div>
  <SearchInputForm
    buttonText="go"
    name="searchInput"
    onSubmit={[Function]}
    placeholder="Type Player's ID"
  />
</div>
`;
